libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp

```diff
@@ -151,6 +311,52 @@ bool AP_BattMonitor_SMBus::read_word(uint8_t reg, uint16_t& data) const
     return true;
 }
 
+// read_block - returns number of characters read if successful, zero if unsuccessful, if append_zero true
+// appends '\0' to the end of the array
+uint8_t AP_BattMonitor_SMBus::read_block(uint8_t reg, uint8_t* data, bool append_zero) const
+{
+    // get length
+    uint8_t bufflen;
+    // read byte (first byte indicates the number of bytes in the block)
+    if (!_dev->read_registers(reg, &bufflen, 1)) {
+        return 0;
+    }
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > SMBUS_READ_BLOCK_MAXIMUM_TRANSFER) {
+        return 0;
+    }
+
+    // buffer to hold results (2 extra byte returned holding length and PEC)
+    const uint8_t read_size = bufflen + 1 + (_pec_supported ? 1 : 0);
```

Comment:
looks like read_size needs to be 16 bit, otherwise is bufflen == 255 we wrap to zero and get stack corruption